File: sdk/python/src/teamusers_sdk/middleware.py
"""Framework-neutral authentication and authorization helpers."""

from __future__ import annotations

from collections.abc import Callable, Mapping
from dataclasses import dataclass
from typing import Any

from .permissions import AllowResult, CheckResult, PermissionsClient
from .verifier import Claims, SDKError, TokenVerificationError, Verifier
from .types import Resource


class UnauthorizedError(SDKError):
    """The request is not authenticated (HTTP 401)."""

    code = "UNAUTHORIZED"
    status_code = 401
    status = 401
    statusCode = 401


class ForbiddenError(SDKError):
    """The authenticated subject lacks a permission (HTTP 403)."""

    code = "FORBIDDEN"
    status_code = 403
    status = 403
    statusCode = 403


AuthenticationError = UnauthorizedError
AuthorizationError = ForbiddenError


def _request_headers(request: Any) -> Mapping[str, Any]:
    if isinstance(request, Mapping):
        headers = request.get("headers", request.get("Headers", request))
    else:
        headers = getattr(request, "headers", getattr(request, "Headers", {}))
    return headers if isinstance(headers, Mapping) else {}


def _authorization_header(headers: Mapping[str, Any]) -> str | None:
    for key, value in headers.items():
        if str(key).lower() == "authorization":
            if isinstance(value, bytes):
                try:
                    value = value.decode("utf-8")
                except UnicodeDecodeError:
                    return None
            return value if isinstance(value, str) else None
    return None


def _bearer_token(header: str | None) -> str | None:
    if not isinstance(header, str):
        return None
    parts = header.split()
    if len(parts) != 2 or parts[0].lower() != "bearer" or not parts[1]:
        return None
    return parts[1]


def Authenticate(request: Any, verifier: Verifier | None = None) -> Claims:
    """Verify a bearer token from a minimal request shape.

    The argument order mirrors the TypeScript/Go middleware helper.  For
    compatibility with early Python callers, ``Authenticate(verifier,
    request)`` is accepted as well.
    """

    if isinstance(request, Verifier) and verifier is not None and not isinstance(verifier, Verifier):
        request, verifier = verifier, request
    if verifier is None:
        raise UnauthorizedError("authentication verifier unavailable")
    token = _bearer_token(_authorization_header(_request_headers(request)))
    if token is None:
        raise UnauthorizedError("bearer token is required")
    try:
        return verifier.verify(token)
    except SDKError as error:
        raise UnauthorizedError("authentication failed", error) from error
    except Exception as error:
        raise UnauthorizedError("authentication failed", error) from error


def authenticate(request: Any, verifier: Verifier | None = None) -> Claims:
    return Authenticate(request, verifier)


def _client_allow(
    client: Any,
    claims: Claims,
    permission: str,
    resource: Resource | Mapping[str, Any] | None,
) -> tuple[bool, str]:
    method = getattr(client, "Allow", None) or getattr(client, "allow", None)
    if not callable(method):
        return False, "authorization client unavailable"
    try:
        result = method(claims, permission, resource)
    except SDKError:
        return False, "authorization service unavailable"
    if isinstance(result, tuple):
        if len(result) >= 2:
            return bool(result[0]), str(result[1])
        if len(result) == 1:
            return bool(result[0]), ""
    if hasattr(result, "allow"):
        return bool(result.allow), str(getattr(result, "reason", ""))
    return bool(result), "permission granted" if result else "permission denied"


def Require(
    client: Any,
    request: Any = None,
    claims: Claims | None = None,
    permission: str | None = None,
    resource: Resource | Mapping[str, Any] | Callable[[Any], Resource | Mapping[str, Any]] | None = None,
):
    """Require one permission, returning claims or raising a typed 403.

    The direct form is ``Require(client, request, claims, permission,
    resource)``.  Calling ``Require(client, permission)`` returns a handler
    accepting ``request`` and optional ``claims`` for middleware composition.
    """

    if isinstance(request, str) and permission is None:
        configured_permission = request
        resource_from = resource if callable(resource) else None

        def handler(next_request: Any, next_claims: Claims | None = None, next_resource: Any = None) -> Claims:
            resolved_resource = next_resource
            if resolved_resource is None and resource_from is not None:
                resolved_resource = resource_from(next_request)
            return Require(client, next_request, next_claims, configured_permission, resolved_resource)

        return handler

    if permission is None or not isinstance(permission, str):
        raise ForbiddenError("permission is required")
    if claims is None:
        verifier = getattr(client, "verifier", getattr(client, "Verifier", None))
        claims = Authenticate(request, verifier)
    allowed, reason = _client_allow(client, claims, permission, resource if not callable(resource) else resource(request))
    if not allowed:
        raise ForbiddenError(reason or "permission denied")
    return claims


def require(
    client: Any,
    request: Any = None,
    claims: Claims | None = None,
    permission: str | None = None,
    resource: Resource | Mapping[str, Any] | Callable[[Any], Resource | Mapping[str, Any]] | None = None,
):
    return Require(client, request, claims, permission, resource)


@dataclass(slots=True)
class Client:
    """Verifier plus optional local/remote permission authorization."""

    verifier: Verifier | None = None
    permissions: PermissionsClient | None = None
    remote_only: bool = False

    @property
    def Verifier(self) -> Verifier | None:
        return self.verifier

    @property
    def Permissions(self) -> PermissionsClient | None:
        return self.permissions

    @property
    def RemoteOnly(self) -> bool:
        return self.remote_only

    def authenticate(self, request: Any) -> Claims:
        return Authenticate(request, self.verifier)

    def Authenticate(self, request: Any) -> Claims:
        return self.authenticate(request)

    def allow(
        self,
        claims: Claims,
        permission: str,
        resource: Resource | Mapping[str, Any] | None = None,
    ) -> AllowResult:
        if self.permissions is None:
            return AllowResult(False, "authorization client unavailable")
        return self.permissions.allow(claims, permission, resource, remote_only=self.remote_only)

    def Allow(
        self,
        claims: Claims,
        permission: str,
        resource: Resource | Mapping[str, Any] | None = None,
    ) -> AllowResult:
        return self.allow(claims, permission, resource)

    def check(
        self,
        subject: str,
        permission: str,
        resource: Resource | Mapping[str, Any] | None = None,
    ) -> CheckResult:
        if self.permissions is None:
            return CheckResult(allow=False, reason="authorization client unavailable")
        return self.permissions.check(subject, permission, resource)

    def Check(
        self,
        subject: str,
        permission: str,
        resource: Resource | Mapping[str, Any] | None = None,
    ) -> CheckResult:
        return self.check(subject, permission, resource)

    def require(self, permission: str, resource_from: Callable[[Any], Any] | None = None):
        return Require(self, permission, resource=resource_from)

    def Require(self, permission: str, resource_from: Callable[[Any], Any] | None = None):
        return self.require(permission, resource_from)

    def middleware(self, request: Any, next_handler: Callable[[Any, Claims], Any] | None = None) -> Any:
        claims = self.authenticate(request)
        if next_handler is None:
            return claims
        return next_handler(request, claims)

    def Middleware(self, request: Any, next_handler: Callable[[Any, Claims], Any] | None = None) -> Any:
        return self.middleware(request, next_handler)

    def SubscribePermissions(self, source_or_url: Any, handler: Callable[[list[str]], Any] | None = None):
        if self.permissions is None:
            raise UnauthorizedError("permission client is unavailable")
        return self.permissions.SubscribePermissions(source_or_url, handler)


NewClient = Client


__all__ = [
    "AuthenticationError",
    "AuthorizationError",
    "Authenticate",
    "Client",
    "ForbiddenError",
    "NewClient",
    "Require",
    "UnauthorizedError",
    "authenticate",
    "require",
]

File: sdk/ts/src/middleware.ts
import {
  SDKError,
  TokenVerificationError,
  type Claims,
  type Verifier,
} from "./verifier.js";
import type { Resource } from "./abac.js";

/** Minimal request shape required by the SDK middleware helpers. */
export interface MiddlewareRequest {
  readonly headers?: Headers | Readonly<Record<string, string | readonly string[] | undefined>>;
  readonly method?: string;
  readonly [key: string]: unknown;
}

export class UnauthorizedError extends SDKError {
  public readonly status = 401;
  public readonly statusCode = 401;

  public constructor(message = "authentication is required", cause?: unknown) {
    super("UNAUTHORIZED", message, cause);
  }
}

export class ForbiddenError extends SDKError {
  public readonly status = 403;
  public readonly statusCode = 403;

  public constructor(message = "permission denied", cause?: unknown) {
    super("FORBIDDEN", message, cause);
  }
}

export interface AuthorizerLike {
  readonly verifier?: Verifier;
  allow(
    claims: Claims,
    permission: string,
    resource?: Resource,
    signal?: AbortSignal,
  ): Promise<{ allow: boolean; reason: string }>;
}

/** Extract and verify a bearer token from a minimal request object. */
export async function authenticateRequest(
  request: MiddlewareRequest,
  verifier: Pick<Verifier, "verify">,
): Promise<Claims> {
  const raw = bearerToken(headerValue(request.headers, "authorization"));
  if (raw === undefined) {
    throw new UnauthorizedError("bearer token is required");
  }
  try {
    return await verifier.verify(raw);
  } catch (error) {
    if (error instanceof UnauthorizedError) throw error;
    throw new UnauthorizedError("authentication failed", error);
  }
}

/** Go-style Authenticate helper; accepts either (request, verifier) or (verifier, request). */
export async function Authenticate(
  first: MiddlewareRequest | Pick<Verifier, "verify">,
  second: MiddlewareRequest | Pick<Verifier, "verify">,
): Promise<Claims> {
  if (isVerifier(first)) {
    return authenticateRequest(second as MiddlewareRequest, first);
  }
  return authenticateRequest(first, second as Pick<Verifier, "verify">);
}

export const authenticate = Authenticate;
export const AuthenticateRequest = authenticateRequest;

/** Create a permission middleware function for a request pipeline. */
export function requirePermission(
  authorizer: AuthorizerLike,
  permission: string,
  resourceFrom?: Resource | ((request: MiddlewareRequest) => Resource | Promise<Resource>),
): (request: MiddlewareRequest, claims?: Claims) => Promise<Claims> {
  return async (request: MiddlewareRequest, providedClaims?: Claims): Promise<Claims> => {
    const claims = providedClaims ?? await authenticateFromAuthorizer(request, authorizer);
    const resource = typeof resourceFrom === "function"
      ? await resourceFrom(request)
      : resourceFrom ?? {};
    let decision: { allow: boolean; reason: string };
    try {
      decision = await authorizer.allow(claims, permission, resource);
    } catch (error) {
      throw new ForbiddenError("authorization failed", error);
    }
    if (!decision.allow) {
      throw new ForbiddenError(decision.reason || "permission denied");
    }
    return claims;
  };
}

/** Go-style Require helper; supports a factory or direct request invocation. */
export function Require(
  authorizer: AuthorizerLike,
  permission: string,
  resourceFrom?: Resource | ((request: MiddlewareRequest) => Resource | Promise<Resource>),
): (request: MiddlewareRequest, claims?: Claims) => Promise<Claims>;
export function Require(
  authorizer: AuthorizerLike,
  request: MiddlewareRequest,
  claims: Claims | undefined,
  permission: string,
  resource?: Resource,
): Promise<Claims>;
export function Require(
  authorizer: AuthorizerLike,
  second: string | MiddlewareRequest,
  third?: Resource | ((request: MiddlewareRequest) => Resource | Promise<Resource>) | Claims,
  fourth?: string,
  fifth?: Resource,
): ((request: MiddlewareRequest, claims?: Claims) => Promise<Claims>) | Promise<Claims> {
  if (typeof second === "string") {
    return requirePermission(authorizer, second, third as Resource | ((request: MiddlewareRequest) => Resource | Promise<Resource>) | undefined);
  }
  if (fourth === undefined) {
    throw new TypeError("permission is required");
  }
  return requirePermission(authorizer, fourth, fifth)(second, third as Claims | undefined);
}

function bearerToken(header: string | undefined): string | undefined {
  if (header === undefined) return undefined;
  const parts = header.trim().split(/\s+/u);
  if (parts.length !== 2 || parts[0].toLowerCase() !== "bearer" || parts[1] === "") return undefined;
  return parts[1];
}

function headerValue(
  headers: Headers | Readonly<Record<string, string | readonly string[] | undefined>> | undefined,
  name: string,
): string | undefined {
  if (headers === undefined) return undefined;
  if (headers instanceof Headers) return headers.get(name) ?? undefined;
  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() !== name) continue;
    if (typeof value === "string") return value;
    if (Array.isArray(value)) return value.join(",");
    return undefined;
  }
  return undefined;
}

async function authenticateFromAuthorizer(
  request: MiddlewareRequest,
  authorizer: AuthorizerLike,
): Promise<Claims> {
  if (authorizer.verifier === undefined) {
    throw new UnauthorizedError("authentication verifier unavailable");
  }
  return authenticateRequest(request, authorizer.verifier);
}

function isVerifier(value: unknown): value is Pick<Verifier, "verify"> {
  return typeof value === "object" && value !== null && "verify" in value && typeof value.verify === "function";
}
export { UnauthorizedError as AuthenticationError };
export { ForbiddenError as AuthorizationError };
export { TokenVerificationError };
